### ISSUE TITLE
Explain the Elm Architecture in "Plain JS" before learning Elm Language?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-elm#82]
The issue of the "Elm Learning Curve" was raised in: https://github.com/dwyl/learn-elm/issues/45
and scrolling down to to @lucymonie's [list](https://github.com/dwyl/learn-elm/issues/45#issuecomment-275947200) we see the Elm Architecture at number four ...

`this` seems fairly logical (_initially_) because the _Elm **Guide**_ 
uses the _Elm **Language**_ to explain the _Elm **Architecture**_: https://guide.elm-lang.org/architecture 
![elm-architecture](https://cloud.githubusercontent.com/assets/194400/25693159/92aca968-309f-11e7-9dbb-69a62aa95907.png)

i.e. it ***assumes*** that people **already _understand_** the (Core) _Elm **Language**_...
This is a _fair_ assumption given the ordering of the Guide _however_... 

## Hypothesis

I ***hypothesize*** that if we _**explain** the **Elm Architecture**_ (_**in** a bit more **detail**_) 
using a **language** people are _**already familiar**_ with (_i.e **JavaScript**_) `before` diving into the Elm Language it will ["flatten"](https://english.stackexchange.com/questions/6212/whats-the-opposite-for-steep-learning-curve) the learning curve.

# tl;dr

## _Analogy_

Let us _briefly_ ["_reason by **analogy**_"](https://youtu.be/NV3sBlRgzTI): Picture yourself learning a `new` (_human/foreign_) language.

> This scenario _assumes_ you have a _**First** (Human) **Language**_ (_e.g: English<sup>1</sup>_) and you are learning another one, let's assume the `new` language is your _second language_. (_you with me so far?_)

Imagine (_in this scenario_) that you have _never_ written an essay before (_in any human language_) 
so you don't understand [**essay structure**](https://writingcenter.fas.harvard.edu/pages/essay-structure). 

The teacher has just spent the lesson going through the basic vocabulary in the `new` language,
and now gives you a homework assignment:  "_write an **essay** using all the words you've learned_!" 

If you have _never_ written an _essay_ before, this _could **easily overwhelm** you_. 

I _posit_ that (_at least for `some` people_) understanding the Architecture 
in a Language they _already_ know (_i.e. JS_) might make it easier.

<sup>1</sup>_If your **first** (human) **Language** is **not english**, substitute for your own._

### Axioms

1. **JavaScript** is (_by far_) the "***most popular***" **programming language** by _any_ measure.
2. **_Most_ people** will not learn Elm as their _first_ (_programming_) language.
3. _Familiarity_ with a subject aids comprehension of topics that _build_ on that knowledge. 
e.g: **Algebra** skills aid in learning _**Physics**_. 
or (_simpler_): knowing how to ride a **bicycle** will help with a ***motorcycle***.

![image](https://cloud.githubusercontent.com/assets/194400/25694372/36438bb8-30a6-11e7-9b65-11374888a997.png)
JavaScript is the most popular programming language on GitHub by _every measure_.
It could be argued that more code exists _off_ GitHub in "Enterprise" (_Closed Source_) repositories,
but we have _no way_ of reliably quantifying it.

![image](https://cloud.githubusercontent.com/assets/194400/25694517/ecfe27a0-30a6-11e7-8722-360d24294938.png)
We _know_ that a _vanishingly_ small number of people use/know `elm` (_evidenced by ["**star count**"](https://github.com/elm-lang/elm-compiler)_), 
the idea that someone would learn `elm` as their _first_ programming language is _unlikely_.
Richard Feldman spends the first _hour_ of his Elm Course comparing elm to JS (ES2015).
This is not "proof" that "_everyone_" learning Elm already knows JavaScript, but it's substantive.

I _propose_ writing a _separate_ tutorial _just_ for the **Elm _Architecture_** in "***Plain JavaScript***".
I made an _attempt_ to distil the Architecture in ES5 JS yesterday:
https://github.com/nelsonic/practice/blob/master/elm.html
But think it deserves a _comprehensive_ treatment ...

**Comment by gabrielperales** [comment]
@nesonic , maybe this repo can help you with that task https://github.com/anler/elm-architecture-cover .

**Comment by nelsonic** [comment]
Morning @gabrielperales, how are you [_today_](https://www.google.co.uk/search?q=May+The+Fourth+Be+With+You&tbm=isch) ? 😉 
https://github.com/anler/elm-architecture-cover and https://github.com/kaleidos/olmo are a _good_ examples however I think we can do _much_ better
by:
a) clearly stating what the assumptions/prerequisites are (_and minimising them!_)
b) smaller steps 
c) even more basic examples
d) tests!!

**Comment by newswim** [comment]
Wonder if you saw this repo with a few nice `.md` overviews -> https://github.com/elm-guides/elm-for-js

**Comment by nelsonic** [comment]
Hi @newswim, yeah the elm-guides are _good_, however they are still _taught_ in Elm.
I am proposing explaining the **Elm _Architecture_** in **(_Plain_) JavaScript** to people who _know_ JavaScript and encouraging them to _build_ a **_JavaScript_ application** _using_ the **Elm _Architecture_**.
I propose that once people understand the **Elm _Architecture_** and have _used_ it in JS,
_that_ part of the "Elm Learning Curve" will be much easier.

**Comment by YvesMuyaBenda** [comment]
@nelsonic Is this going to amount to functional reactive programming done in Javascript, translating the concepts piece by piece into the corresponding Javascript concepts?

**Comment by nelsonic** [comment]
JavaScript's dominance _cemented_ for the foreseeable future:
http://www.stanforddaily.com/2017/02/28/cs-department-updates-introductory-courses

**Comment by nelsonic** [comment]
@YvesMuyaBenda the HtDP approach https://en.wikipedia.org/wiki/How_to_Design_Programs
would be a _good_ one. 
My _plan_ is to build up the concepts of the Elm Architecture using TDD. 
Will ping you when there's _something_ coherent. 😉

**Comment by YvesMuyaBenda** [comment]
@nelsonic Looking forward to the ping man! Looking forward to the ping! It is is a bit difficult going from the initial idea to something coherent, no? But that is where the fun lies, the stretching out of one's mind.

**Comment by nelsonic** [comment]
Done: https://github.com/dwyl/learn-elm-architecture-in-javascript
@lucymonie, @YvesMuyaBenda, @newswim _feedback_ very much appreciated!

## Are the tutorial steps ***clear enough*** for ***complete beginners***?

`CC`: @Cleop, @rjmk, @jrans, @jruts, @katbow, @gabrielperales, @benjaminlees, @shouston3, @SimonLab, @andrewMacmurray, @Danwhy, @jackcarlisle, @naazy, @Conorc1000, @roryc89

**Comment by newswim** [comment]
This is FANTASTIC, good work @nelsonic !!
